### ISSUE TITLE
Fish fails assertion on commandline: `fish<tab>` (complete, complet-and-search)

### ISSUE THREAD

**Issue by da99** [fish-shell/fish-shell#4179]
I have `bind \t complete` (default binding). I've also tried the following with `bind \t complete-and-search`. In both cases, fish crashes (fails: assert(desc_remainging >= 2) in src/pager.cpp:181) when commandline == `fish`. For any other term it works:

![fish 1](https://user-images.githubusercontent.com/1551848/27778734-ed722a5a-5f82-11e7-9d6c-7507a6a243e6.png)

It works for other terms (e.g. nvi, fi, ech, ..). (Although, sometimes it works with `fish`, but then fails on `fi`.)
![fish 2](https://user-images.githubusercontent.com/1551848/27778735-ed774062-5f82-11e7-99b5-7abdd49bec5c.png)

I'm guessing this is due to my setup? 

fish 2.6.0
The problem occurs on various libvte-based terminals in an X session
and outside of X in the default Linux VT terminal.
Void Linux, using the XBPS package manager.

**Comment by krader1961** [comment]
Assertion failures like you reported in this issue should never occur. Nonetheless there is something extremely unusual about your environment. Can you share your fish config? That is, the contents of *~/.config/fish* and its subdirectories? Also, what does `locale` output?

How are your launching the X session? That appears to be the essential aspect of this issue given your original problem statement.

**Comment by da99** [comment]
Well the problem occurs outside of X, on a bare Linux login terminal. I've tried changing default login shells (sh, bash, zsh) using `chsh` and the problem persists.  The problem also persists even after removing my custom ~/.config/fish setup, on a brand new setup. 

The output of `locale` is:
```sh
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
```

**Comment by krader1961** [comment]
I don't understand why you think changing your default login shell would make a difference. Are you saying that if you make your default login shell `bash` then login using that shell and type `fish` you see this problem?

This seems suspiciously like an earlier problem I fixed wherein fish did not correctly handle terminals that incorrectly reported its size. What does `stty -a` output before you run fish? What does `echo $COLUMNS $LINES` output?

I am also confused by your original problem statement where you say "... when commandline == fish. For any other term it works." I appreciate that English may not be your native language but that statement doesn't make sense to a native speaker of English.

**Comment by da99** [comment]
Right... I wanted to be sure if the login shell was doing something strange... just to rule out the possibility.

I've installed fish on another computer, (same OS and architecture, VoidLinux 64-bit AMD), and it ran fine with the default config. So I guess the problem is with this particular machine. I checked the packages and compilation options. I haven't been able to find any differences between the two machines. 
 
![fish 4](https://user-images.githubusercontent.com/1551848/27779736-9e9bb574-5f8a-11e7-92a2-c7221d19ba4d.png)

```
speed 38400 baud; rows 24; columns 80; line = 0;
intr = ^C; quit = ^\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;
eol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;
werase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;
-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts
-ignbrk brkint ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl -ixon -ixoff
-iuclc -ixany imaxbel iutf8
opost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0
isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
echoctl echoke -flusho -extproc

```

**Comment by da99** [comment]
"... when commandline == fish. For any other term it works."
I meant: When I press the kets `fish<tab>` on the fish command line, the error usually occurs.
Other times, when I press the keys `fi<tab>`,  the problem occurs.

However, if fish fails the assertion on `fi<tab>`, then typing `fish<tab>` works. But when `fish<tab>` fails, then typing `fi<tab>` works. ("Works" in this case means that the "complete menu" is shown under the command line.)

Also, when I type other letters that are the beginnings of other programs on the installed paths (e.g. n-v-i, then press <tab>, the menu is show underneath. For example: `nvi<tab>` shows the menu that includes the names of executable such as `nvi`, `nvim`, etc.

(I apologize for my poor writing. It is indeed shameful that I have been speaking English for 20+ years.)  I also don't want to waste any more of your time. Would you like me to close this issue? Since no one else has encountered it, it will probably go away in a few package upgrades in the coming weeks.

**Comment by zanchey** [comment]
Is it the first Tab press, or do you have to press it more than once?

**Comment by da99** [comment]
@zanchey It's the first tab press.

**Comment by krader1961** [comment]
I'd prefer not to close this issue because, as I said earlier, assertion failures should never happen.

Looking at the code around the `assert()` failure it is almost a certainty the problem is the description for one of the completion items has a width fish is incorrectly calculating. Most likely because it is not valid UTF-8 or contains characters whose width is ambiguous or fish is incorrectly calculating. The question is how to validate that hypothesis.

Please reproduce the failure then type

```
complete --do-complete=fi > /tmp/x
```

Replace `fi` with `fish` or whatever you had typed when the failure occurred. Then attach */tmp/x* to this issue. It might be helpful to do this on the machine exhibiting the problem and what should be an identical machine that does not exhibit the problem.

**Comment by da99** [comment]
Attached is the files.
The first is when it fails, the second is when it does not fail. 

[x.fail.txt](https://github.com/fish-shell/fish-shell/files/1120927/x.fail.txt)
[x.pass.txt](https://github.com/fish-shell/fish-shell/files/1120928/x.pass.txt)

**Comment by krader1961** [comment]
Thanks, @da99. Unfortunately there isn't anything interesting in those files. Which doesn't surprise me since it appears that `complete --do-complete` does not include the same descriptions you see when actually pressing [tab].

Are you able to apply a patch to the fish source and build and run a custom fish? That is likely to provide us with useful information to get to the root cause of this problem. The alternative is to have you run `set -c unlimited` then send us the resulting core file. Unfortunately there is a pretty low probability we'll be able to make use of the core dump.

**Comment by da99** [comment]
I'll use the patch option then since that is the most useful.

**Comment by krader1961** [comment]
Fantastic, @da99. I'll try to create a patch in the next couple of days that will hopefully provide some insight into why you're seeing that assertion failure.

**Comment by da99** [comment]
@krader1961 Out of curiosity, are you on a Linux distro? BSD? MacOS?

**Comment by krader1961** [comment]
@da99, All of the above. The systems that I use for this type of interaction run macOS but I regularly use GNU LInux and BSD as backend systems.

**Comment by faho** [comment]
@da99: One of Void's selling points is that it offers a choice of libc - glibc or musl. Which one are you using?

I've just tried the glibc edition and couldn't reproduce this.

**Comment by faho** [comment]
I finally got the musl edition working (the graphical ones just black out in any virtualization solution I've tried, so I had to fallback to the X-less one), and I can't reproduce it either.

So it most likely is connected to your system. One possibility is that you have an executable with unusual characters in $PATH. Mind posting the contents of that here? `printf '%s\n' $PATH/*` should work.

**Comment by da99** [comment]
@faho Both of the PCs I use are glibc (VoidLinux)

Attached is the output of `printf '%s\n' $PATH/* >/tmp/paths`. The first is on the machine that fails. The second is on the machine that passes the assertion.

[fails.txt](https://github.com/fish-shell/fish-shell/files/1122754/fails.txt)
[passes.txt](https://github.com/fish-shell/fish-shell/files/1122753/passes.txt)

**Comment by da99** [comment]
@faho Would you like me to try to re-install every package on the system? Maybe the problem is due to a package or library that is not properly installed?

**Comment by krader1961** [comment]
Please download the attached patch. Then do the following in the fish source directory:

```
> patch -p1 < patch.txt
> ./configure --prefix=/usr/local
> make fish
> ./fish
```

If you then reproduce the problem several lines of diagnostic data will be written to your screen. Copy/paste it into this issue.

[patch.txt](https://github.com/fish-shell/fish-shell/files/1122792/patch.txt)

**Comment by da99** [comment]
(I ran `autoreconf --no-recursive` after applying the patch to generate the `configure` script.)
Here is the output:

[fish.patched.txt](https://github.com/fish-shell/fish-shell/files/1122829/fish.patched.txt)

```
WTF orig_desc_remaining 9
WTF desc_remaining 1
WTF c->comp[0] 11 |_key_reader|
WTF c->desc 1809 |- explore what characters keyboard keys send Synopsis fish_key_reader [OPTIONS] Description fish_key_reader is used to study input received from the terminal and can help with key binds. The program is interactive and works on standard input. Individual characters themselves and their hexadecimal values are displayed. The tool will write an example bind command matching the character sequence captured to stdout. If the character sequence matches a special key name (see bind --key-names), both bind CHARS ... and bind -k KEYNAME ... usage will be shown. Additional details about the characters received, such as the delay between chars, are written to stderr. The following options are available: o 2 -c or --continuous begins a session where multiple key sequences can be inspected. By default the program exits after capturing a single key sequence. o 2 -d or --debug-level=DEBUG_LEVEL enables debug output and specifies a verbosity level (like fish -d). Defaults to 0. o 2 -D or --debug-stack-frames=DEBUG_LEVEL specify how many stack frames to display when debug messages are written. The default is zero. A value of 3 or 4 is usually sufficient to gain insight into how a given debug call was reached but you can specify a value up to 128. o 2 -h or --help prints usage information. Usage Notes The delay in milliseconds since the previous character was received is included in the diagnostic information written to stderr. This information may be useful to determine the optimal fish_escape_delay_ms setting or learn the amount of lag introduced by tools like ssh, mosh or tmux. fish_key_reader intentionally disables handling of many signals. To terminate fish_key_reader in --continuous mode do: o 2 press Ctrl-C twice, or o 2 press Ctrl-D twice, or o 2 type exit, or o 2 type quit|
WTF c->comp_width 15
WTF c->desc_width 18446744073709551615
<E> fish: src/pager.cpp:193: failed assertion: desc_remaining >= 2
<E> fish: Backtrace:
<E> fish: 0   pager_t::completion_print_item(std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, pager_t::comp_t const*, unsigned long, unsigned long, unsigned long, bool, bool, page_rendering_t*) const + 1476
<E> fish: 1   pager_t::completion_print(unsigned long, unsigned long const*, unsigned long, unsigned long, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<pager_t::comp_t, std::allocator<pager_t::comp_t> > const&, page_rendering_t*) const + 465
<E> fish: 2   pager_t::completion_try_print(unsigned long, std::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<pager_t::comp_t, std::allocator<pager_t::comp_t> > const&, page_rendering_t*, unsigned long) const + 532
<E> fish: 3   pager_t::render() const + 469
<E> fish: 4   pager_t::update_rendering(page_rendering_t*) const + 44
<E> fish: 5   ./fish() [0x4c1033]
<E> fish: 6   reader_readline(int) + 1607
<E> fish: 7   reader_read(int, io_chain_t const&) + 540
<E> fish: 8   main + 3031
<E> fish: 9   __libc_start_main + 241
<E> fish: 10  _start + 42
zsh: abort      ./fish
```

**Comment by da99** [comment]
@krader1961 Something unusual I found with this patched fish build:
The assertion fails with the window width is 1075 or higher. (X session, running BSPwm).
But passes when the window width is 1070 or below.
And it fails in the default Linux login virtual terminal, outside of X. (So I don't think X is the problem.)
Also, height does not seem to matter in this case.

In the installed, pre-compiled version of the package manager, the width was irrelevant in failing the assertion. The installed, pre-compiled package used CXX flags:  `-D_GNU_SOURCE=1 -D_ISO99_SOURCE=1`

**Comment by da99** [comment]
@krader1961 Here is the same info. above using columns instead of window widths:
For the patched fish build:
When the terminal is 134 columns or higher, assertion fails.
When its 133 or below, assertion passes.

**Comment by krader1961** [comment]
The completion description width is 0xffffffffffffffff. Which is -1 if treated as a signed int. So much for size_t helping avoid this type of bug.

@da99, This doesn't have anything to do with your terminal width. The failure is because `fish_wcswidth()` is returning -1 for that description. But assigning -1 to an unsigned int converts it to the largest value an unsigned int can hold. When fish then adds 2 to that value the result is 1. Which screws up this logic:

```
         while (desc_remaining > c->desc_width + 2) {
             desc_remaining -= print_max(L" ", punct_color, 1, false, &line_data);
         }
```

That loop is supposed to terminate when `desc_remaining` is a minimum of 4. But because of the aforementioned problem it terminates with `desc_remaining` set to 2.

Since the description appears to consist entirely of ASCII characters it isn't obvious why `fish_wcswidth()` would return -1 (indicating it couldn't calculate the width).

**Comment by faho** [comment]
@da99: Can you upload the output of `__fish_describe_command fish_key_reader` and `apropos fish_key_reader`?

**Comment by da99** [comment]
@faho 
`__fish_describe_command fish_key_reader`: 
[__fish_describe.txt](https://github.com/fish-shell/fish-shell/files/1122938/__fish_describe.txt)

`apropos fish_key_reader`: 
[apropos.txt](https://github.com/fish-shell/fish-shell/files/1122939/apropos.txt)

**Comment by krader1961** [comment]
The description contains backspace characters. Which don't have a width thus causing `fish_wcswidth()` to return -1. TBD is the best solution. Although at a minimum we need to check if we're assigning -1 to the completion description width and substitute zero.

**Comment by krader1961** [comment]
Also, what does `apropos fish_key_reader` output on the system where you don't see this problem? On my macOS and Ubuntu systems I get much shorter description:

```
fish_key_reader (1)  - explore what characters keyboard keys send -
```

**Comment by da99** [comment]
@krader1961  On the system w/o the problem: 
```
~> apropos fish_key_reader
apropos: nothing appropriate
```

**Comment by da99** [comment]
@krader1961 Also, both systems have `man-pages` installed as a package. 
However, the system with the problem does not have `man-db` installed.

**Comment by faho** [comment]
@da99: That's why it's not happening. Your install is missing fish_key_reader's man page (and possibly the command itself). That system is broken differently.

@krader1961: See https://github.com/fish-shell/fish-shell/issues/4179#issuecomment-312939859. I already asked about apropos 😄 - `__fish_describe_command` doesn't cope with that output.

We should fix that (though it's an awk script, and I'm not great with that), then we should also trim everything after e.g. 30 characters in case it fails for some _other_ apropos, or in case the man page is too wordy. Descriptions are supposed to be short. (That alone would avoid the issue in this particular case, but of course it's not a real fix)

Then we should probably do something about pager.cpp.

**Comment by da99** [comment]
@faho "That system is broken differently." Thanks. For some strange reason, that cheered me up. :smile: 

Just some minor notes: `fish_key_reader` is installed on both systems @ /usr/bin/fish_key_reader. The output of `type fish_key_reader` on both systems:
```
fish_key_reader is a function with definition
# Defined in - @ line 0
function fish_key_reader --description 'alias fish_key_reader=/usr/bin/fish_key_reader'
    /usr/bin/fish_key_reader $argv

end
```

**Comment by krader1961** [comment]
Out of curiosity, why are you doing `alias fish_key_reader=/usr/bin/fish_key_reader `?

**Comment by faho** [comment]
>"That system is broken differently." Thanks. For some strange reason, that cheered me up. 😄

Hehe... I just missed your message - the missing mandb is the cause. Without that, `apropos` won't find anything, so the description won't be generated, so no naughty characters will reach our poor pager.

**Comment by da99** [comment]
@krader1961 Not sure if this is important, but on both my systems, I'm getting:
`fish_key_reader: unrecognized option '--version'` . Which then causes `fish_key_reader` to be aliased in this file:

cat share/functions/fish_key_reader.fish
```fish
# check if command fish_key_reader works and is the same version that
# came with this fish. This will happen one time.
command -sq fish_key_reader
and command fish_key_reader --version 2>&1 | string match -rq $FISH_VERSION
# if alias doesn't define the function here, this is an autoloaded "nothing".
# the command (if there is one) will be used by default.
or alias fish_key_reader=(string escape $__fish_bin_dir/fish_key_reader)

```

**Comment by krader1961** [comment]
@da99, Thanks for bringing that to our attention as well as the pager.cpp bug. That script has been broken since last November since the fkr program has never had a `--version` flag.

**Comment by faho** [comment]
@da99: No, it's not hugely important, but it is a bit unclean.

That function file was added for .app bundles (and similar) according to 95385eda80ec2fbabd05ee0f01fac42a70cf6dad and db5eebd37240adb5fed785a39b79585a78489c8d, and it checks the version because of 4c34ba995985e08b947246dca62caa90315d9523.

I've added the `--version` flag now, so it should no longer be happening.

**Comment by da99** [comment]
@krader1961 and @faho Thanks for all your help on this. Glad my broken systems were good for something :smiley:

**Comment by faho** [comment]
Thanks for helping make fish better! This bug could probably also be hit with other completions, so it's actually significant.

**Comment by da99** [comment]
Maintaining the pager seems like very difficult work because of all the ways a system can be customized/broken. 

How practical/impractical would it be to replace it with an external program? 

Some seem very feature-rich and well maintained. (One example: `fzf`. It's one of the more customizable ones out there and runs on OSX/Windows/Linux.)

**Comment by faho** [comment]
>Maintaining the pager (i.e. menu selection dialog for completion and suggestions) seems like very difficult work because of all the ways a system can be customized/broken.

I'm sure it was hard work when it was written, but that's mostly done now. There really aren't that many pager bugs - in fact this is the only one I can recall.

There also isn't that much customization related to the pager itself.

>How practical/impractical would it be to replace it with an external program?

It's possible. In fact, here's a PoC implementation for using fzf:

```fish
function fzf_complete
    set -l cmdline (commandline)
    # HACK: Color descriptions manually.
    complete -C | string replace -r \t'(.*)$' \t(set_color $fish_pager_color_description)'$1'(set_color normal) \
    | fzf -d \t -1 -0 --ansi --header="$cmdline" --tabstop=4 \
    | read -l token
    # Remove description
    set token (string replace -r \t'.*' '' -- $token)
    commandline -rt "$token"
end
```

(This can be extended with e.g. `--height=X%`, and pick a single candidate immediately. I've mostly written it just to see if it was possible)

Fzf also ships with a few fish functions, including ones for picking files, commands from history, directories to cd to.

If you want, you can customize your install to pretty much ignore the pager.

However, we don't want this by default because:

- It adds another dependency - this makes it harder to install because you now also need to build fzf

- It integrates a bit worse - it won't respect fish's color settings

- It behaves a bit differently - I _like_ the current pager behavior for the most part

- It means giving up control - it would mean letting the fzf developers pick the behavior, and waiting for fzf releases for fixes and features.

- It's not necessary - the current pager is already done and works well, so why change?

**Comment by da99** [comment]
@faho You're completely right. Thanks again for writing such a complete answer. It also helped me understand more about fish's design. It seems to give me everything I want without the need to read 1000 man pages.

**Comment by heinrich5991** [comment]
I'm running into this issue as well. The current development version fixes the problem for me. What is your release schedule, when can we expect a new release?

**Comment by heinrich5991** [comment]
I have **no** idea how I unassigned someone from this issue – I shouldn't even have the rights to do so...

**Comment by zanchey** [comment]
2.7b1 is already overdue - I am working on it.

**Comment by mqudsi** [comment]
Maybe this should go out in a 2.6.1?

**Comment by heinrich5991** [comment]
Apparently this is a Github bug, and me commenting just triggered a cleanup which in turn unassigned a person that is no longer a member of this repository.
